### ISSUE TITLE
New overflow predicates

### ISSUE THREAD

**Issue by LeventErkok** [LeventErkok/sbv#657]
SMTLib has adapted new overflow predicates, and z3 implements them:

```
bvnego
bvuaddo bvsaddo
bvusubo bvssubo
bvumulo bvsmulo
bvsdivo
```

(Not sure if z3 implements the `bvumolo` and `bvsmulo`, need to check.)

We should make sure SBV uses these sanctioned versions when they come out in the new SMTLib document. (Which isn't out yet.)

**Comment by LeventErkok** [comment]
Waiting till there's an official SMTLib Bitvector logic update. As of mid-2023 this hasn't happened yet.

**Comment by RyanGlScott** [comment]
For my own curiosity, what constitutes an "SMTLib Bitvector logic update"? I see that these five of these operations (`bvnego`, `bvuaddo`, `bvsaddo`, `bvumulo`, and `bvsmulo`) are now listed in https://smt-lib.org/theories-FixedSizeBitVectors.shtml.

**Comment by LeventErkok** [comment]
I'm not sure if I understand your question. It just means Cesare (or other SMTLib folks) updated the documentation that you linked to.

Of course, this only makes real sense if solvers (z3/cvc5 etc) do implement the update, but I think they're careful in not pushing an update until they at least implement and experiment with it in one solver for these sorts of things. (SMTLib3 is a whole another can of worms.)

Is that what you are asking about? I feel like I'm missing something.

**Comment by RyanGlScott** [comment]
Ah, gotcha! The documentation was updated to include these operations a while ago (2024-07-16), so I was wondering if there was another reason to hold back using these in SBV.

I think most popular SMT solvers (notably, `bitwuzla`, `cvc5`, and `z3`) now support these operations. One notable exception is `yices-smt2`, which seems a bit behind in supporting newer SMT-LIB operations (see, for instance, https://github.com/SRI-CSL/yices2/issues/359#issuecomment-5143954269).

**Comment by LeventErkok** [comment]
indeed. I think I found bitwuzla to have the best overflow implementation in my experiments, but this was a while ago.

I'm not sure what the status of Yices is, really. Since Bruno moved from SRI, I think it's kind of on a "low-maintenance" mode? I wouldn't hold my breath for it to add support. (And, to be honest, I haven't used yices in a very long time. I wouldn't be surprised lots of SBV things didn't work with it.)

**Comment by RyanGlScott** [comment]
To be clear, I don't have a particular need to support this feature in Yices. I was mainly wondering if SBV could add support these operations in the solvers that do support them. These operations are pretty handy!

(I found this while perusing which newer SMT-LIB operations `what4` does not yet support. Another interesting one I found is [integer exponentation](https://smt-lib.org/theories-Ints.shtml) (e.g., `(** x y)`). Unfortunately, I haven't found _any_ solver which supports that one yet.)

**Comment by LeventErkok** [comment]
hold on; we do support them in SBV, don't we?  I thought we did!

**Comment by LeventErkok** [comment]
See `Data.SBV.Tools.Overflow`

Not sure these are exported from Dynamic though. Perhaps that's what you meant?

**Comment by LeventErkok** [comment]
Exponentiation is interesting. I didn't realize they had that. (I think SBV supports it if the second argument is a constant, but clearly that's not good enough.) Can you double check and file a GitHub issue if not?  Solver support is important, but if it's in the spec, we should support it.

**Comment by RyanGlScott** [comment]
Apologies, I completely misinterpreted the comment in https://github.com/LeventErkok/sbv/issues/657#issuecomment-1581027867 as "I'm closing the issue preemptively until the SMT-LIB documentation is opened", but in reality, this was fixed quite ago in https://github.com/LeventErkok/sbv/commit/5a8ea1b7763ea2b68b09bdc2d921e77aa38d434b. (I blame GitHub's search---it didn't come up with any results for `bvnego` when I tried several minutes ago, but it does come up with search results now. Ugh.)

In any case, thank you for the response (and sorry for the noise)!
